### PR TITLE
Fix/soc 6028 : Exception in the logs when assigning a user as a manager

### PR DESCRIPTION
**Pull request by abeji**
The NullPointerException is due to the fact that the logged-on user is not registered in the post save user event.
The solution is so to add this information by setting the Editor value on the space object before the event broadcast.

### CODE REVIEW

File: component/core/src/main/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImpl.java [hunk 1]
```diff
@@ -58,7 +58,7 @@ public void postDelete(Membership m) throws Exception {
       Space space = spaceService.getSpaceByGroupId(m.getGroupId());
 
       ConversationState state = ConversationState.getCurrent();
-      if(state != null) {
+      if(space != null && state != null) {
```
Comment (thomasdelhomenie) [inline component/core/src/main/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImpl.java]: We should do the same check on state.getIdentity() to make sure it is not null, like you did in the postSave method.

File: component/core/src/main/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImpl.java [hunk 2]
```diff
@@ -88,6 +88,11 @@ public void postSave(Membership m, boolean isNew) throws Exception {
       Space space = spaceService.getSpaceByGroupId(m.getGroupId());
       //TODO A case to confirm: will we create a new space here when a new group is created via organization portlet
       if (space != null) {
+        //Set Editor value on the space object
```
Comment (thomasdelhomenie) [inline component/core/src/main/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImpl.java]: I think this comment is useless.

File: component/core/src/main/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImpl.java [hunk 3]
```diff
@@ -88,6 +88,11 @@ public void postSave(Membership m, boolean isNew) throws Exception {
...
+        ConversationState state = ConversationState.getCurrent();
+        if(state != null && state.getIdentity() != null) {
+          space.setEditor(state.getIdentity().getUserId());
```
Comment (thomasdelhomenie) [inline component/core/src/main/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImpl.java]: Shouldn't we check if an editor already exist in the space object before setting it ?

File: component/core/src/main/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImpl.java (hunk 2 above)
Comment (abeji) [reply component/core/src/main/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImpl.java]: Done

File: component/core/src/main/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImpl.java (hunk 3 above)
Comment (abeji) [reply component/core/src/main/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImpl.java]: Done

File: component/core/src/main/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImpl.java (hunk 1 above)
Comment (abeji) [reply component/core/src/main/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImpl.java]: Done

Comment (thomasdelhomenie) [reply component/core/src/main/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImpl.java]: That's better but we should move the check on (space != null) higher IMO. If the space is null, it will throw a NPE some lines above where space is used.

**Comment by thomasdelhomenie** [comment]
@abeji please update unit tests also.

File: component/core/src/main/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImpl.java (hunk 1 above)
Comment (thomasdelhomenie) [reply component/core/src/main/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImpl.java]: Not in the case space is null.

Comment (abeji) [reply component/core/src/main/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImpl.java]: Done

**Comment by abeji** [comment]
Unit tests have been added

File: component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java [hunk 4]
```diff
@@ -65,9 +65,13 @@ public void testPostSave() {
...
     sFoo.setRegistration(Space.OPEN);
     sFoo = createSpaceNonInitApps(sFoo, "root", null);
-
+    String editor = sFoo.getEditor();
     addUserToGroupWithMembership("demo", sFoo.getGroupId(), "manager");
     sFoo = spaceService.getSpaceById(sFoo.getId());
+    if(editor == null)
+      assertEquals("root", sFoo.getEditor());
+    else
+      assertEquals(editor, sFoo.getEditor());
```
Comment (thomasdelhomenie) [inline component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java]: Tests should not contain if/else. It is up to you to set or not the editor and check the behavior in each with assertions. It probably requires to add test methods for each case.

Comment (abeji) [reply component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java]: Done

File: component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java [hunk 5]
```diff
@@ -87,6 +83,12 @@ public void testPostSave() {
     sFoo = spaceService.getSpaceById(sFoo.getId());
     assertEquals(3, sFoo.getManagers().length);
     assertEquals(4, sFoo.getMembers().length);
+    
+    //update space Editor when it's null
+    sFoo.setEditor(null);
+    addUserToGroupWithMembership("john", sFoo.getGroupId(), "manager");
+    sFoo = spaceService.getSpaceById(sFoo.getId());
+    assertEquals(sFoo.getEditor(),"root");
```
Comment (thomasdelhomenie) [inline component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java]:
```suggestion
    assertEquals(sFoo.getEditor(), "root");
```

File: component/service/src/main/java/org/exoplatform/social/rest/impl/comment/CommentRestResourcesV1.java [hunk 6]
```diff
@@ -74,7 +76,7 @@ public Response getCommentById(@Context UriInfo uriInfo,
                                  @ApiParam(value = "Comment id", required = true) @PathParam("id") String id,
                                  @ApiParam(value = "Asking for a full representation of a specific subresource if any", required = false) @QueryParam("expand") String expand) throws Exception {
 
-    ExoSocialActivity act = activityManager.getActivity(id);
+    ExoSocialActivity act = activityManager.getActivity(COMMENT_PREFIX + id);
```
Comment (thomasdelhomenie) [inline component/service/src/main/java/org/exoplatform/social/rest/impl/comment/CommentRestResourcesV1.java]: Is this change related to the space editor issue ?

File: component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java [hunk 7]
```diff
@@ -65,9 +60,10 @@ public void testPostSave() {
     sFoo.setType(DefaultSpaceApplicationHandler.NAME);
     sFoo.setRegistration(Space.OPEN);
     sFoo = createSpaceNonInitApps(sFoo, "root", null);
-
+    sFoo.setEditor("another user");
     addUserToGroupWithMembership("demo", sFoo.getGroupId(), "manager");
     sFoo = spaceService.getSpaceById(sFoo.getId());
+    assertEquals(sFoo.getEditor(),"another user");
```
Comment (thomasdelhomenie) [inline component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java]: We should a valid username IMO, spaces are not allowed in usernames. You can use one of the created user, for example "mary".

File: component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java [hunk 8]
```diff
@@ -118,20 +120,27 @@ public void testPostDelete() throws Exception {
...
-    sFoo = spaceService.getSpaceById(sFoo.getId());
     assertEquals(1, sFoo.getManagers().length);
     assertEquals(2, sFoo.getMembers().length);
+    
+    //update space Editor when it's null
+    addUserToGroupWithMembership("john", sFoo.getGroupId(), "manager");
+    sFoo.setEditor(null);
+    membershipHandler.removeMembershipByUser("john", true);
+    sFoo = spaceService.getSpaceById(sFoo.getId());
+    assertEquals(sFoo.getEditor(),"root");
```
Comment (thomasdelhomenie) [inline component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java]:
```suggestion
    assertEquals(sFoo.getEditor(), "root");
```

File: component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java (hunk 7 above)
Comment (abeji) [reply component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java]: Done

File: component/service/src/main/java/org/exoplatform/social/rest/impl/comment/CommentRestResourcesV1.java (hunk 6 above)
Comment (abeji) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/comment/CommentRestResourcesV1.java]: No it's related to another issue, I revert this changes.

File: component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java [hunk 9]
```diff
@@ -118,20 +120,27 @@ public void testPostDelete() throws Exception {
     assertEquals(1, sFoo.getManagers().length);
     assertEquals(3, sFoo.getMembers().length);
     //
+    sFoo.setEditor("mary");
     membershipHandler.removeMembershipByUser("john", true);
     sFoo = spaceService.getSpaceById(sFoo.getId());
+    assertEquals(sFoo.getEditor(),"mary");
```
Comment (thomasdelhomenie) [inline component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java]:
```suggestion
    assertEquals(sFoo.getEditor(), "mary");
```

File: component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java [hunk 10]
```diff
@@ -65,9 +60,10 @@ public void testPostSave() {
     sFoo.setType(DefaultSpaceApplicationHandler.NAME);
     sFoo.setRegistration(Space.OPEN);
     sFoo = createSpaceNonInitApps(sFoo, "root", null);
-
+    sFoo.setEditor("mary");
     addUserToGroupWithMembership("demo", sFoo.getGroupId(), "manager");
     sFoo = spaceService.getSpaceById(sFoo.getId());
+    assertEquals(sFoo.getEditor(),"mary");
```
Comment (thomasdelhomenie) [inline component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java]:
```suggestion
    assertEquals(sFoo.getEditor(), "mary");
```

File: component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java (hunk 9 above)
Comment (abeji) [reply component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java]: Done

File: component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java (hunk 10 above)
Comment (abeji) [reply component/core/src/test/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImplTest.java]: Done
